modules/gdscript/gdscript_parser.cpp

```diff
@@ -6255,7 +6322,9 @@ bool GDScriptParser::_is_type_compatible(const DataType &p_container, const Data
 	switch (p_expression.kind) {
 		case DataType::NATIVE: {
 			if (p_container.kind != DataType::NATIVE) {
-				// Non-native type can't be a superclass of a native type
+				if (p_container.kind == DataType::GDSCRIPT || p_container.kind == DataType::SCRIPT) {
+					return ScriptServer::get_global_class_name(p_container.script_type->get_path()) == p_expression.native_type;
+				}
```

Comment:
1. Wouldn't this check need to be made the other way around? Here, you are checking to see if the container is a script and the expression is a native class, e.g. `var my_node: MyNode = Node.new()`. However, the scripted type would have to extend the native type, so you would be assigning a base type to a derived type member. But that would *break* polymorphism, allowing you to assign incompatible values to variables.

1. Even if you move it, you still need to get what the base type is and *then* query the ClassDB for inheritance with `is_parent_class()` rather than making a direct comparison to the base type. After all, the script could extend a class *deriving* the member's expected type, e.g.

        var a_control: Control = MyTextureRect.new()

    where `MyTextureRect` is a script extending `TextureRect` rather than `Control`.

Reply:
Yes, this code doesn't seem to be right... but the whole type checking in 3.2 GDscript is a big mess. The type checking goes through so many code paths it's hard to keep track of everything.

I did some tests and my changes don't really affect any of the cases you mentioned (I still think my code is incorrect, but it doesn't really make things worse in these cases). Comparing This PR and 3.2.2 stable I get the exact same behavior in this test script:
![gdscript_errors](https://user-images.githubusercontent.com/4402304/90518870-966f9e80-e167-11ea-91c6-3c0b4fc46fe3.png)

So... I'm not really sure what is the best course of action here. I hope with the cleanup of GDscript 2.0 we get a type checker that is much easier to maintain because the 3.2 one is a bit of a nightmare.